### PR TITLE
Remove year 0 ticks if generated

### PR DESCRIPTION
**Pull request by jbusecke**
Towards fixing #44. 

@spencerkclark I was not sure how to add tests for this. Is there an example that you could point me to?

### CODE REVIEW

**Comment by spencerkclark** [comment]
Disclaimer: I'm not a maintainer of this package so take my suggestion with a grain of salt :).  @trexfeathers was kind enough to review my recent (closed) PR; perhaps he'd be willing to weigh in here.  

My two cents: if maintainers agree with this fix, I think adding some unit tests of `NetCDFTimeDateLocator.tick_values` that check this new behavior would make sense.  You might be able to draw some inspiration from the existing test cases of this method:

https://github.com/SciTools/nc-time-axis/blob/531dd0da83f2ea6b351f129559ccd4e8b0ccbb5f/nc_time_axis/tests/unit/test_NetCDFTimeDateLocator.py#L81-L115

**Comment by jbusecke** [comment]
Ah sorry. I didn't do my research. Just associated you with the package. I'll take a look in a bit.

**Comment by jbusecke** [comment]
@trexfeathers are you generally ok with this fix? It is not a really sleek solution, but this problem has been popping up in many of my workflows: I work with long control runs of climate models which start in year 1, but the logic tends to try and 'pad' the axis limits a little bit and this cause a tick at year 0.

**Comment by trexfeathers** [comment]
>@trexfeathers are you generally ok with this fix?

@jbusecke I only looked at nc-time-axis for the first time two weeks ago 😆, so I'd be uncomfortable passing judgement without doing some background reading first. I'm hopeful that I or another in my team will be permitted the time to give this the attention it deserves next week.

**Comment by trexfeathers** [comment]
@jbusecke it is also important that you sign the SciTools CLA (see the [Governance](https://scitools.org.uk/organisation.html#governance) notes for more detail), otherwise this PR cannot be merged even after it has been reviewed.

**Comment by jbusecke** [comment]
Thanks for the instructions. Just signed the form.

**Review by trexfeathers — CHANGES_REQUESTED** [review]
OK @jbusecke, I've stared and prodded at this for a while and I'm confident that your methodology is the right one - the Matplotlib API for `mticker.MaxNLocator()` is such that there's no good way of forecasting what will be generated, especially with all the specific `resolution` handling that nc_time_axis does, so it's definitely best to go for a retrospective modification as you have.

---

As for testing, I suggest:
* Copying [`Test_tick_values.test_yearly()`](https://github.com/SciTools/nc-time-axis/blob/9bbc2f709537c6a581e7353e83845dc7a7b5c8ec/nc_time_axis/tests/unit/test_NetCDFTimeDateLocator.py#L113-L115), to make a new test
  (Obviously if one of the other existing tests makes for a better starting template then feel free to use whatever is appropriate)
* Adapt the new test to generate an example of what you are trying to fix, and assert for a range of ticks that does **not** include the tick that your code should remove
* Give the new test a name that is descriptive of the scenario it is testing for

---

I've also commented with a specific suggested change to your proposed code. Thanks!

**Comment by jbusecke** [comment]
Hi @trexfeathers, super sorry for the long delay. 

I have added a test, but I am quite unsure about the format (have only used pytest so far). My basic idea was to create ticks that would default to year 0 and loop over every calendar type. I also checked that none of the ticks has the `.year` 0, but it is probably even ok to just confirm that the test does not fail. 

Happy to change this any way you see fit.

File: nc_time_axis/tests/unit/test_NetCDFTimeDateLocator.py [hunk 1]
```diff
@@ -115,5 +115,43 @@ def test_yearly(self):
             self.check(5, 0, 5*365), [31., 485., 942., 1399., 1856.])
 
 
+class Test_tick_values_yr0(unittest.TestCase):
```
Comment (trexfeathers) [inline nc_time_axis/tests/unit/test_NetCDFTimeDateLocator.py:118]: Given this is quite different from the others, I agree it deserves its own class 👌

File: nc_time_axis/tests/unit/test_NetCDFTimeDateLocator.py [hunk 2]
```diff
@@ -115,5 +115,43 @@ def test_yearly(self):
...
+                                        date_unit=self.date_unit)
+        return locator.tick_values(num1, num2)
+
+    def test_yearly_yr0_remove(self):
+            yr0_remove_calendars = [
+                "proleptic_gregorian",
+                "gregorian",
+                "julian",
+                "standard"
+                    ]
```
Comment (trexfeathers) [inline nc_time_axis/tests/unit/test_NetCDFTimeDateLocator.py:134]: Move to `setUp` - it is likely that `test_yearly_yr0_remove()` will remain the only test in `Test_tick_values_yr0`, but even if we end up with other tests they will also be interested in this object.

Also valuable since we want the actual test to have the minimum possible code outside the behaviour under test.

File: nc_time_axis/tests/unit/test_NetCDFTimeDateLocator.py [hunk 3]
```diff
@@ -115,5 +115,43 @@ def test_yearly(self):
...
+                'standard',
+                'gregorian',
+                'proleptic_gregorian',
+                'noleap',
+                '365_day',
+                '360_day',
+                'julian',
+                'all_leap',
+                '366_day',
+                    ]:
```
Comment (trexfeathers) [inline nc_time_axis/tests/unit/test_NetCDFTimeDateLocator.py:146]: Create this list in `setUp` - it is likely that `test_yearly_yr0_remove()` will remain the only test in `Test_tick_values_yr0`, but even if we end up with other tests they will also be interested in this object.

Also valuable since we want the actual test to have the minimum possible code outside the behaviour under test.

File: nc_time_axis/tests/unit/test_NetCDFTimeDateLocator.py [hunk 4]
```diff
@@ -115,5 +115,43 @@ def test_yearly(self):
...
+
+                # convert values to dates, check that none of them has year 0
+                ticks = self.check(5, 0, 100*365, calendar)
+                if calendar in yr0_remove_calendars:
```
Comment (trexfeathers) [inline nc_time_axis/tests/unit/test_NetCDFTimeDateLocator.py:150]: It's important to also assert that `0` still exists for the other calendars in an `else:` - confirming that you're actually making a difference and it's not just passing my accident.

File: nc_time_axis/tests/unit/test_NetCDFTimeDateLocator.py [hunk 5]
```diff
@@ -115,5 +115,43 @@ def test_yearly(self):
...
+                    for t in ticks:
+                        assert cftime.utime(
+                            self.date_unit,
+                            calendar).num2date(t).year != 0
```
Comment (trexfeathers) [inline nc_time_axis/tests/unit/test_NetCDFTimeDateLocator.py:154]: We should be able to get this a bit neater, without a loop, by using assignment and list comprehension.  List comprehension is usually more efficient, and in my opinion the below is a little more readable too. What do you think of this?

(I've written this as a pure re-write, not accounting for my other suggestions)

```suggestion
num2date = cftime.utime(self.date_unit, calendar).num2date
year_ticks = [num2date(t).year for t in ticks]
if calendar in yr0_remove_calendars:
    self.assertNotIn(0, year_ticks)
```

**Review by trexfeathers — CHANGES_REQUESTED** [review]
@jbusecke thanks for persevering, especially having to use unittest 🤢 - we all want pytest too, when there's time!

And sorry for the gradual pace of this; I'm regularly pushing for more time to look at nc-time-axis.

I reckon you're nearly there, but have some suggestions for improved coverage and good setup practice.

**Comment by trexfeathers** [comment]
> but it is probably even ok to just confirm that the test does not fail.

I had a similar thought, but the few extra lines we are working on make it much clearer what we're aiming for.

File: nc_time_axis/tests/unit/test_NetCDFTimeDateLocator.py (hunk 5 above)
Comment (jbusecke) [reply nc_time_axis/tests/unit/test_NetCDFTimeDateLocator.py]: Yes that is great. The initial one was kind of a mess, since I am not very familiar with the internal structure of the package. Thanks a lot for the suggestions.

File: nc_time_axis/tests/unit/test_NetCDFTimeDateLocator.py (hunk 4 above)
Comment (jbusecke) [reply nc_time_axis/tests/unit/test_NetCDFTimeDateLocator.py]: I added another line to address this. Let me know if you think this is sufficient.

File: nc_time_axis/tests/unit/test_NetCDFTimeDateLocator.py (hunk 3 above)
Comment (jbusecke) [reply nc_time_axis/tests/unit/test_NetCDFTimeDateLocator.py]: Done

**Comment by jbusecke** [comment]
Ok I think I have implemented all the changes you requested (and reversed the automatic formatting introduced by my editor). 
Let me know if there is anything else you would want me to change. 

Many thanks for the patient review!

**Review by trexfeathers — APPROVED** [review]
Nice job, LGTM!
